cdap-ranger/cdap-ranger-binding/src/main/java/co/cask/cdap/security/authorization/ranger/binding/RangerAuthorizer.java

```diff
@@ -0,0 +1,306 @@
+    Properties properties = context.getExtensionProperties();
+    instanceName = properties.containsKey(Constants.INSTANCE_NAME) ?
+      properties.getProperty(Constants.INSTANCE_NAME) : "cdap";
+    if (rangerPlugin == null) {
+      try {
+        UserGroupInformation ugi = UserGroupInformation.getLoginUser();
+        if (ugi != null) {
+          // set the login user as the user as whom cdap is running as this is needed for kerberos authentication
+          MiscUtil.setUGILoginUser(ugi, null);
+        }
+        LOG.debug("Initializing Ranger CDAP Plugin with UGI {}", ugi);
+      } catch (Throwable t) {
+        LOG.error("Error getting principal.", t);
+      }
+      // the string name here should not be changed as this uniquely identifies the service in ranger. If it's
+      // changed it will require changing all the supporting xml file which is in this package.
+      rangerPlugin = new RangerBasePlugin("cdap", "cdap");
+    }
+    rangerPlugin.init();
+    RangerDefaultAuditHandler auditHandler = new RangerDefaultAuditHandler();
+    rangerPlugin.setResultProcessor(auditHandler);
+  }
+
+
+  @Override
+  public void enforce(EntityId entity, Principal principal, Action action) throws Exception {
+    LOG.debug("Enforce called on entity {}, principal {}, action {}", entity, principal, action);
+    if (rangerPlugin == null) {
+      throw new RuntimeException("CDAP Ranger Authorizer is not initialized.");
+    }
+
+    if (principal.getType() != Principal.PrincipalType.USER) {
+      throw new IllegalArgumentException(String.format("The principal type for current enforcement request is '%s'. " +
+                                                         "Authorization enforcement is only supported for '%s'.",
+                                                       principal.getType(), Principal.PrincipalType.USER));
+    }
+    String requestingUser = principal.getName();
+    String ip = InetAddress.getLocalHost().getHostName();
+    java.util.Set<String> userGroups = MiscUtil.getGroupsForRequestUser(requestingUser);
+
+    LOG.debug("Requesting user {}, ip {}, requesting user groups {}", requestingUser, ip, userGroups);
+
+    Date eventTime = new Date();
+    String accessType = toRangerAccessType(action);
+    RangerAccessRequestImpl rangerRequest = new RangerAccessRequestImpl();
+    rangerRequest.setUser(requestingUser);
+    rangerRequest.setUserGroups(userGroups);
+    rangerRequest.setClientIPAddress(ip);
+    rangerRequest.setAccessTime(eventTime);
+
+    RangerAccessResourceImpl rangerResource = new RangerAccessResourceImpl();
+    rangerRequest.setResource(rangerResource);
+    rangerRequest.setAccessType(accessType);
+    rangerRequest.setAction(accessType);
+    rangerRequest.setRequestData(entity.toString());
+
+    setAccessResource(entity, rangerResource);
+
+    boolean isAuthorized = false;
+
+    try {
+      RangerAccessResult result = rangerPlugin.isAccessAllowed(rangerRequest);
+      if (result == null) {
```

Comment:
It means that authorization enforcement didn't happen so ranger cannot tell you whether the access is allowed or not. Now it's up to the called how to handle it whether throw exception deny authorization.